xds/src/test/java/io/grpc/xds/XdsClientFallbackTest.java

```diff
@@ -0,0 +1,319 @@
+    xdsClient.watchXdsResource(XdsRouteConfigureResource.getInstance(), RDS_NAME, rdsWatcher);
+
+    restartServer(TdServerType.MAIN);
+
+    inOrder.verify(ldsWatcher, timeout(10000)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(MAIN_HTTP_CONNECTION_MANAGER));
+  }
+
+  // Since everything is loaded we shouldn't fallback
+  @Test
+  public void connect_then_mainServerDown_fallbackServerUp()  throws Exception {
+    restartServer(TdServerType.MAIN);
+    restartServer(TdServerType.FALLBACK);
+    xdsClient = xdsClientPool.getObject();
+
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(), MAIN_SERVER, ldsWatcher);
+
+    verify(ldsWatcher, timeout(10000)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(MAIN_HTTP_CONNECTION_MANAGER));
+
+    mainTdServer.getServer().shutdownNow();
+
+    // Shouldn't do fallback since all watchers are loaded
+    verify(ldsWatcher, timeout(5000).times(0)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(FALLBACK_HTTP_CONNECTION_MANAGER));
+
+    // Should just get from cache
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(), MAIN_SERVER, ldsWatcher2);
+    xdsClient.watchXdsResource(XdsRouteConfigureResource.getInstance(), RDS_NAME, rdsWatcher);
+    verify(ldsWatcher2, timeout(10000)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(MAIN_HTTP_CONNECTION_MANAGER));
+    verify(ldsWatcher, never()).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(FALLBACK_HTTP_CONNECTION_MANAGER));
+
+    // Asking for something not in cache should force a fallback
+    xdsClient.watchXdsResource(XdsClusterResource.getInstance(), "cluster0", cdsWatcher);
+    verify(rdsWatcher, timeout(10000)).onChanged(any());
+    verify(ldsWatcher, timeout(10000)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(FALLBACK_HTTP_CONNECTION_MANAGER));
+    verify(ldsWatcher2, timeout(10000)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(FALLBACK_HTTP_CONNECTION_MANAGER));
+    verify(cdsWatcher, timeout(10000)).onChanged(any());
+  }
+
+  @Test
+  public void connect_then_mainServerRestart_fallbackServerdown()  throws Exception {
+    restartServer(TdServerType.MAIN);
+    xdsClient = xdsClientPool.getObject();
+
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(), MAIN_SERVER, ldsWatcher);
+
+    verify(ldsWatcher, timeout(10000)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(MAIN_HTTP_CONNECTION_MANAGER));
+
+    mainTdServer.getServer().shutdownNow();
+    fallbackServer.getServer().shutdownNow();
+
+    xdsClient.watchXdsResource(XdsClusterResource.getInstance(), "cluster0", cdsWatcher);
+    restartServer(TdServerType.MAIN);
+    verify(cdsWatcher, timeout(10000)).onChanged(any());
+    verify(ldsWatcher, timeout(10000).times(2)).onChanged(
+        XdsListenerResource.LdsUpdate.forApiListener(MAIN_HTTP_CONNECTION_MANAGER));
+  }
+
```

Comment:
I don't see the test coverage of verifying "Once such a resource is received, the XdsClient will close connections to the lower-priority server(s) "

Reply:
Done